src/main.ts

```diff
@@ -13,88 +12,74 @@ import GitHubRelease, {
   IGitHubReleaseOptions
 } from './github-release';
 
+import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
 import init from './init';
+import NpmPlugin from './plugins/npm/npm';
+import SEMVER from './semver';
 import execPromise from './utils/exec-promise';
+import getGitHubToken from './utils/github-token';
 import createLog from './utils/logger';
 
-const readFile = promisify(fs.readFile);
+interface IAuthor {
+  name: string;
+  email: string;
+}
+
+interface IRepository {
+  owner: string;
+  repo: string;
+  token?: string;
+}
 
-const calcGreaterVersion = (
-  packageVersion: string,
-  monorepoVersion: string,
-  latestVersion: string
-) => {
-  const localVersion = monorepoVersion || packageVersion;
+export interface IAutoHooks {
+  beforeRun: SyncHook<[IGitHubReleaseOptions]>;
+  getAuthor: AsyncSeriesBailHook<[], IAuthor>;
+  getPreviousVersion: AsyncSeriesBailHook<
+    [(release: string) => string],
+    string
+  >;
+  getRepository: AsyncSeriesBailHook<[], IRepository>;
+  publish: AsyncSeriesHook<[SEMVER]>;
+}
 
-  return localVersion > latestVersion ? localVersion : latestVersion;
-};
+async function getRepo(args: ArgsType, hooks: IAutoHooks) {
```

Comment:
Rather than passing around the `hooks` object everywhere, it might make sense for us to create an `AutoRelease` class, put the hooks on that, and then add all of these functions as methods on the `AutoRelease` class. It falls more in line with how we've used hooks previously ( a `hooks` property on a class) -- and it's nicer from an API standpoint:


```js
const autoRelease = new AutoRelease({ 
    plugins: [ new NpmPlugin() ]
});

autoRelease.run()
```